esphome/components/msa3xx/__init__.py

```diff
@@ -0,0 +1,186 @@
+MSA_RANGES = {
+    "2G": MSARange.RANGE_2G,
+    "4G": MSARange.RANGE_4G,
+    "8G": MSARange.RANGE_8G,
+    "16G": MSARange.RANGE_16G,
+}
+
+MSAResolution = msa3xx_ns.enum("Resolution", True)
+MSA_RESOLUTIONS = {
+    14: MSAResolution.RES_14BIT,
+    12: MSAResolution.RES_12BIT,
+    10: MSAResolution.RES_10BIT,
+    8: MSAResolution.RES_8BIT,
+}
+
+_COMMON_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(MSA3xxComponent),
+        cv.Optional(CONF_RANGE, default="2G"): cv.enum(MSA_RANGES, upper=True),
+        cv.Optional(CONF_CALIBRATION): cv.Schema(
+            {
+                cv.Optional(CONF_OFFSET_X, default=0): cv.float_range(
+                    min=-4.5, max=4.5
+                ),
+                cv.Optional(CONF_OFFSET_Y, default=0): cv.float_range(
+                    min=-4.5, max=4.5
+                ),
+                cv.Optional(CONF_OFFSET_Z, default=0): cv.float_range(
+                    min=-4.5, max=4.5
+                ),
+            }
+        ),
+        cv.Optional(CONF_TRANSFORM): cv.Schema(
+            {
+                cv.Optional(CONF_MIRROR_X, default=False): cv.boolean,
+                cv.Optional(CONF_MIRROR_Y, default=False): cv.boolean,
+                cv.Optional(CONF_MIRROR_Z, default=False): cv.boolean,
+                cv.Optional(CONF_SWAP_XY, default=False): cv.boolean,
+            }
+        ),
+        cv.Optional(CONF_ON_ACTIVE): automation.validate_automation(single=True),
+        cv.Optional(CONF_ON_TAP): automation.validate_automation(single=True),
+        cv.Optional(CONF_ON_DOUBLE_TAP): automation.validate_automation(single=True),
+        cv.Optional(CONF_ON_FREEFALL): automation.validate_automation(single=True),
+        cv.Optional(CONF_ON_ORIENTATION): automation.validate_automation(single=True),
+    }
+).extend(cv.polling_component_schema("10s"))
+
+
+CONFIG_SCHEMA = cv.typed_schema(
+    {
+        MODEL_MSA301: _COMMON_SCHEMA.extend(
+            {
+                cv.Optional(CONF_RESOLUTION, default=14): cv.enum(MSA_RESOLUTIONS),
+            }
+        ).extend(i2c.i2c_device_schema(0x26)),
+        MODEL_MSA311: _COMMON_SCHEMA.extend(
+            {
+                cv.Optional(CONF_RESOLUTION): cv.invalid(
+                    "MSA311 doesn't support different resolutions"
+                ),
+            }
+        ).extend(i2c.i2c_device_schema(0x62)),
+    },
+    upper=True,
+    enum=MSA_MODELS,
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if config[CONF_TYPE] == MODEL_MSA311:
+        config[CONF_RESOLUTION] = 12
+
+    cg.add(var.set_model(config[CONF_TYPE]))
+    cg.add(var.set_range(MSA_RANGES[config[CONF_RANGE]]))
+    cg.add(var.set_resolution(MSA_RESOLUTIONS[config[CONF_RESOLUTION]]))
+
+    if CONF_TRANSFORM in config:
+        transform = config[CONF_TRANSFORM]
```

Comment:
```suggestion
    if transform := config.get(CONF_TRANSFORM):
```

This is the preferred style for handling config options that may not be present. Apply this to the subsequent code too where possible. Just beware of options that can have values that are Falsy like 0 - in that case it is necessary to write `if (xxx := config.get(CONF_XXX) is not None:`

Reply:
good, i forgot about this codestyle. applied!